File: src/test/java/network/crypta/support/OutputThrottleTest.java
package network.crypta.support;

import java.lang.reflect.Field;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;

import static java.util.concurrent.TimeUnit.MILLISECONDS;
import static java.util.concurrent.TimeUnit.NANOSECONDS;
import static java.util.concurrent.TimeUnit.SECONDS;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

@SuppressWarnings("java:S100")
class OutputThrottleTest {
  private static final long ONE_SECOND_NANOS = SECONDS.toNanos(1);
  private static final String TIME_LAST_TICK_FIELD = "timeLastTick";

  @ParameterizedTest
  @ValueSource(longs = {0L, -1L})
  void constructor_whenNanosPerTickIsNotPositive_expectIllegalArgumentException(long nanosPerTick) {
    // Arrange + Act + Assert
    assertThrows(IllegalArgumentException.class, () -> new OutputThrottle(10L, nanosPerTick, 0L));
  }

  @ParameterizedTest
  @ValueSource(longs = {0L, -1L})
  void constructor_whenMaxTokensIsNotPositive_expectIllegalArgumentException(long maxTokens) {
    // Arrange + Act + Assert
    assertThrows(
        IllegalArgumentException.class, () -> new OutputThrottle(maxTokens, ONE_SECOND_NANOS, 0L));
  }

  @Test
  void constructor_whenInitialTokensExceedMax_expectCountClampedToMax() {
    // Arrange
    OutputThrottle throttle = new OutputThrottle(5L, ONE_SECOND_NANOS, 99L);

    // Act
    long count = throttle.getCount();

    // Assert
    assertEquals(5L, count);
  }

  @Test
  void getNanosPerTick_whenConstructed_expectConfiguredValue() {
    // Arrange
    long nanosPerTick = 42_000L;
    OutputThrottle throttle = new OutputThrottle(10L, nanosPerTick, 0L);

    // Act
    long actual = throttle.getNanosPerTick();

    // Assert
    assertEquals(nanosPerTick, actual);
  }

  @ParameterizedTest
  @ValueSource(longs = {0L, -1L})
  void changeNanosAndBucketSize_whenNanosPerTickIsNotPositive_expectIllegalArgumentException(
      long nanosPerTick) {
    // Arrange
    OutputThrottle throttle = new OutputThrottle(10L, ONE_SECOND_NANOS, 0L);

    // Act + Assert
    assertThrows(
        IllegalArgumentException.class, () -> throttle.changeNanosAndBucketSize(nanosPerTick, 10L));
  }

  @ParameterizedTest
  @ValueSource(longs = {0L, -1L})
  void changeNanosAndBucketSize_whenNewMaxIsNotPositive_expectIllegalArgumentException(
      long newMaxTokens) {
    // Arrange
    OutputThrottle throttle = new OutputThrottle(10L, ONE_SECOND_NANOS, 0L);

    // Act + Assert
    assertThrows(
        IllegalArgumentException.class,
        () -> throttle.changeNanosAndBucketSize(ONE_SECOND_NANOS, newMaxTokens));
  }

  @Test
  void changeNanosAndBucketSize_whenNewMaxBelowCurrent_expectCountClippedToNewMax() {
    // Arrange
    OutputThrottle throttle = new OutputThrottle(10L, ONE_SECOND_NANOS, 8L);

    // Act
    throttle.changeNanosAndBucketSize(ONE_SECOND_NANOS, 5L);

    // Assert
    assertEquals(5L, throttle.getCount());
  }

  @Test
  void changeNanosAndBucketSize_whenUpdated_expectNewNanosPerTickReported() {
    // Arrange
    OutputThrottle throttle = new OutputThrottle(10L, ONE_SECOND_NANOS, 1L);
    long newNanosPerTick = 250_000L;

    // Act
    throttle.changeNanosAndBucketSize(newNanosPerTick, 20L);

    // Assert
    assertEquals(newNanosPerTick, throttle.getNanosPerTick());
  }

  @Test
  void forceGrab_whenNegativeTokenCount_expectIllegalArgumentException() {
    // Arrange
    OutputThrottle throttle = new OutputThrottle(10L, ONE_SECOND_NANOS, 0L);

    // Act + Assert
    assertThrows(IllegalArgumentException.class, () -> throttle.forceGrab(-1L));
  }

  @Test
  void forceGrab_whenMoreTokensThanCurrent_expectCountCanBecomeNegative() {
    // Arrange
    OutputThrottle throttle = new OutputThrottle(10L, ONE_SECOND_NANOS, 3L);

    // Act
    throttle.forceGrab(5L);

    // Assert
    assertEquals(-2L, throttle.getCount());
  }

  @Test
  void getCount_whenOneTickElapsed_expectSingleTokenAdded() {
    // Arrange
    OutputThrottle throttle = new OutputThrottle(10L, ONE_SECOND_NANOS, 0L);
    setTimeLastTick(throttle, nowNs() - ONE_SECOND_NANOS);

    // Act
    long count = throttle.getCount();

    // Assert
    assertEquals(1L, count);
  }

  @Test
  void getCount_whenMultipleTicksElapsed_expectMultipleTokensAdded() {
    // Arrange
    OutputThrottle throttle = new OutputThrottle(10L, ONE_SECOND_NANOS, 0L);
    setTimeLastTick(throttle, nowNs() - (4L * ONE_SECOND_NANOS));

    // Act
    long count = throttle.getCount();

    // Assert
    assertEquals(3L, count);
  }

  @Test
  void getCount_whenAccruedTokensExceedMax_expectCountClippedToMax() {
    // Arrange
    OutputThrottle throttle = new OutputThrottle(2L, ONE_SECOND_NANOS, 0L);
    setTimeLastTick(throttle, nowNs() - (10L * ONE_SECOND_NANOS));

    // Act
    long count = throttle.getCount();

    // Assert
    assertEquals(2L, count);
  }

  @Test
  void getCount_whenClockSkewDetected_expectCountUnchangedAndTimestampResynced() {
    // Arrange
    OutputThrottle throttle = new OutputThrottle(10L, ONE_SECOND_NANOS, 4L);
    setTimeLastTick(throttle, nowNs() + (2L * ONE_SECOND_NANOS));

    // Act
    long count = throttle.getCount();
    long updatedTimeLastTick = readTimeLastTick(throttle);

    // Assert
    assertEquals(4L, count);
    assertTrue(updatedTimeLastTick <= nowNs());
  }

  private static long nowNs() {
    long nowMs = System.currentTimeMillis();
    return NANOSECONDS.convert(nowMs, MILLISECONDS);
  }

  private static void setTimeLastTick(OutputThrottle throttle, long value) {
    setLongField(throttle, value);
  }

  private static long readTimeLastTick(OutputThrottle throttle) {
    return readLongField(throttle);
  }

  private static void setLongField(OutputThrottle throttle, long value) {
    try {
      Field field = OutputThrottle.class.getDeclaredField(TIME_LAST_TICK_FIELD);
      field.setAccessible(true);
      field.setLong(throttle, value);
    } catch (ReflectiveOperationException e) {
      throw new IllegalStateException("Failed to set test field: " + TIME_LAST_TICK_FIELD, e);
    }
  }

  private static long readLongField(OutputThrottle throttle) {
    try {
      Field field = OutputThrottle.class.getDeclaredField(TIME_LAST_TICK_FIELD);
      field.setAccessible(true);
      return field.getLong(throttle);
    } catch (ReflectiveOperationException e) {
      throw new IllegalStateException("Failed to read test field: " + TIME_LAST_TICK_FIELD, e);
    }
  }
}

File: src/test/kotlin/network/crypta/support/OutputThrottleTest.kt
package network.crypta.support

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.CsvSource

/**
 * Tests for [OutputThrottle].
 *
 * Design notes for determinism:
 * - Uses a very large `nanosPerTick` so that no automatic token accrual happens during the test
 *   run.
 * - Avoids sleeping and time-based assertions.
 */
@Suppress("kotlin:S100")
class OutputThrottleTest {

  private val oneDayNanos: Long = 86_400_000_000_000L // 24h in nanoseconds

  @Test
  @DisplayName("getCount_whenConstructed_returnsInitialTokensWithoutTicking")
  fun getCount_whenConstructed_returnsInitialTokensWithoutTicking() {
    // Arrange
    val throttle = OutputThrottle(maxTokens = 100, nanosPerTick = oneDayNanos, initialTokens = 42)

    // Act
    val count = throttle.getCount()

    // Assert
    assertEquals(42, count)
  }

  @Test
  @DisplayName("forceGrab_whenEnoughTokens_decrementsBalance")
  fun forceGrab_whenEnoughTokens_decrementsBalance() {
    // Arrange
    val throttle = OutputThrottle(maxTokens = 100, nanosPerTick = oneDayNanos, initialTokens = 100)

    // Act
    throttle.forceGrab(10)
    val remaining = throttle.getCount()

    // Assert
    assertEquals(90, remaining)
  }

  @Test
  @DisplayName("forceGrab_whenInsufficientTokens_allowsNegativeBalance")
  fun forceGrab_whenInsufficientTokens_allowsNegativeBalance() {
    // Arrange
    val throttle = OutputThrottle(maxTokens = 100, nanosPerTick = oneDayNanos, initialTokens = 5)

    // Act
    throttle.forceGrab(10)
    val remaining = throttle.getCount()

    // Assert
    assertEquals(-5, remaining)
  }

  @Test
  @DisplayName("forceGrab_whenNegativeTokens_throws")
  fun forceGrab_whenNegativeTokens_throws() {
    // Arrange
    val throttle = OutputThrottle(maxTokens = 100, nanosPerTick = oneDayNanos, initialTokens = 50)

    // Act + Assert
    assertThrows(IllegalArgumentException::class.java) { throttle.forceGrab(-1) }
  }

  @Test
  @DisplayName("changeNanosAndBucketSize_whenShrinkingBucket_clampsCurrentAndUpdatesNanos")
  fun changeNanosAndBucketSize_whenShrinkingBucket_clampsCurrentAndUpdatesNanos() {
    // Arrange (no ticking due to oneDayNanos)
    val throttle = OutputThrottle(maxTokens = 100, nanosPerTick = oneDayNanos, initialTokens = 90)

    // Act
    throttle.changeNanosAndBucketSize(nanosPerTick = 12345L, newMaxTokens = 50)
    val countAfter = throttle.getCount()
    val nanos = throttle.getNanosPerTick()

    // Assert
    assertEquals(50, countAfter, "Current should be clamped to new max")
    assertEquals(12345L, nanos, "nanosPerTick should be updated")
  }

  @Test
  @DisplayName("getNanosPerTick_whenConstructed_returnsInitialValue")
  fun getNanosPerTick_whenConstructed_returnsInitialValue() {
    // Arrange
    val throttle = OutputThrottle(maxTokens = 10, nanosPerTick = 777L, initialTokens = 0)

    // Act
    val nanos = throttle.getNanosPerTick()

    // Assert
    assertEquals(777L, nanos)
  }

  @ParameterizedTest(name = "ctor invalid args: nanosPerTick={0}, maxTokens={1}")
  @CsvSource(
    // nanosPerTick <= 0
    "0, 10",
    "-1, 10",
    // maxTokens <= 0
    "100, 0",
    "100, -5",
    // both invalid
    "0, 0",
    "-10, -10",
  )
  @DisplayName("constructor_whenInvalidArguments_throws")
  fun constructor_whenInvalidArguments_throws(nanosPerTick: Long, maxTokens: Long) {
    // Arrange + Act + Assert
    assertThrows(IllegalArgumentException::class.java) {
      OutputThrottle(maxTokens = maxTokens, nanosPerTick = nanosPerTick, initialTokens = 0)
    }
  }
}
